File: Iverson.Clients/Java/client/src/main/java/io/iverson/client/search/Query.java
package io.iverson.client.search;

/**
 * Entry point for the Iverson search DSL.
 *
 * <pre>{@code
 * SearchRequest req = Query.of(Article.class)
 *     .where("category").eq("sports")
 *     .orderBy("publishedAt")
 *     .limit(20)
 *     .build();
 * }</pre>
 */
public final class Query {

    private Query() {}

    /**
     * Creates a {@link QueryBuilder} scoped to the given entity class.
     * The type name is derived from {@code entityClass.getSimpleName()}.
     */
    public static <T> QueryBuilder<T> of(Class<T> entityClass) {
        return new QueryBuilder<>(entityClass.getSimpleName());
    }

    /**
     * Creates a {@link QueryBuilder} scoped to the given type name string.
     * Use this when working with the type name directly.
     */
    public static <T> QueryBuilder<T> ofType(String typeName) {
        return new QueryBuilder<>(typeName);
    }

    /**
     * Creates a {@link GroupByBuilder} scoped to the given type name string.
     * Use for compound GROUP BY queries (multiple metrics, HAVING, joins) via
     * {@link GroupByBuilder#build()}.
     */
    public static GroupByBuilder groupBy(String typeName) {
        return new GroupByBuilder(typeName);
    }

    /**
     * Creates a {@link PipelineBuilder} scoped to the given type name. Pipelines compile
     * fluent step chains into one server-side CTE-chain query.
     */
    public static PipelineBuilder pipeline(String typeName) {
        return new PipelineBuilder(typeName);
    }

    /**
     * Creates a {@link SimilarBuilder} for Qdrant vector similarity search on the given
     * embedded property.
     */
    public static SimilarBuilder similar(String typeName, String property) {
        return new SimilarBuilder(typeName, property);
    }

    /**
     * Creates a {@link ChunksBuilder} for Qdrant chunk/RAG search on the given chunked property.
     */
    public static ChunksBuilder chunks(String typeName, String property) {
        return new ChunksBuilder(typeName, property);
    }
}

File: Iverson.Clients/DotNet/Iverson.Client.Search/Query.cs
using System.Linq.Expressions;

namespace Iverson.Client.Search;

/// <summary>
/// Entry point for the Iverson search DSL.
/// Usage: <c>Query.For&lt;Article&gt;().Where(...).OrderBy(...).Page(0).Build()</c>
/// </summary>
public static class Query
{
    public static QueryBuilder<T> For<T>() where T : class => new(typeof(T).Name);

    /// <summary>
    /// Entry point for the GROUP BY DSL. Not generic on a single type, since joins bring
    /// multiple registered types into scope.
    /// </summary>
    public static GroupByBuilder GroupBy(string typeName) => new(typeName);

    /// <summary>
    /// Entry point for the pipeline (CTE chain) DSL. String-based like GroupBy —
    /// steps and joins bring multiple sources into scope. Equivalent to
    /// <see cref="Iverson.Client.Search.Pipeline.For(string)"/>.
    /// </summary>
    public static PipelineBuilder Pipeline(string typeName) => new(typeName);

    /// <summary>
    /// Entry point for Qdrant vector similarity search on a property annotated with
    /// <c>[IversonEmbedding]</c>.
    /// </summary>
    public static QuerySimilarBuilder<T> Similar<T>(Expression<Func<T, object>> property) where T : class =>
        new(typeof(T).Name, PropertyNameObj(property));

    /// <summary>
    /// Entry point for Qdrant chunk/RAG search on a property annotated with <c>[IversonChunk]</c>.
    /// </summary>
    public static QueryChunksBuilder<T> Chunks<T>(Expression<Func<T, object>> property) where T : class =>
        new(typeof(T).Name, PropertyNameObj(property));

    private static string PropertyNameObj<TSource>(Expression<Func<TSource, object>> expr) =>
        expr.Body switch
        {
            MemberExpression member => member.Member.Name,
            UnaryExpression { NodeType: ExpressionType.Convert or ExpressionType.ConvertChecked } unary
                when unary.Operand is MemberExpression innerMember => innerMember.Member.Name,
            _ => throw new ArgumentException("Expression must be a direct property access, e.g. x => x.Title")
        };
}
